### PR TITLE
Improve api

### LINKED ISSUE #20: Flashcard service: Improve API
## Description
Improve the flashcard service API to make it more consistent with the quiz service api.

## Acceptance Criteria
- [ ] API is improved

### PR DESCRIPTION
**Pull request by myluki2000**
## Description of changes
* Improves the API by moving mutations into their own type
* Generally improves the codebase of the service
* Also changes the database structure. A migration will have to be performed
* Includes more thorough automated tests

  
## How has this been tested?

Please describe the test strategy you followed. If not tested, please explain why.

- [x] automated unit test
- [ ] automated integration test
- [ ] manual, exploratory test
    
## Checklist before requesting a review

- [x] My code is easy to understand
- [x] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [x] My code fulfills all acceptance criteria
- [x] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation or
  the [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I made no breaking changes in the database schema or if so, I will perform a database migration

## Checklist for reviewer

- The code is easy to understand
- The code follows
  the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of
  this project
- The code is tested or if not, the reason is documented or discussed
- The added and existing tests reasonably cover the code change
- The code has no breaking changes in the database schema or if so, the assignee is aware of it and
  will perform a database migration

### CODE REVIEW

**Comment by myluki2000** [comment]
DB migration might be a bit more complicated on this one, depending on how exactly hibernate constructs relationships between Entities depending on how you declare your Entity classes in Java.

Is deleting the whole flashcard DB a valid migration strategy? 😅

**Comment by PaulBredl** [comment]
> DB migration might be a bit more complicated on this one, depending on how exactly hibernate constructs relationships between Entities depending on how you declare your Entity classes in Java.
> 
> Is deleting the whole flashcard DB a valid migration strategy? 😅

In this case I would say yes

File: src/main/resources/graphql/query.graphqls [hunk 1]
```diff
@@ -5,8 +5,7 @@ type Query {
     flashcardsByIds(ids: [UUID!]!): [Flashcard!]!
     """
     Get flashcard sets by their assessment ids.
-    Returns a list of lists of flashcard sets, where each list corresponds to a single assessment id.
+    Returns a list of flashcard sets in the same order as the provided ids.
     """
     flashcardSetsByAssessmentIds(assessmentIds: [UUID!]!): [FlashcardSet!]!
```
Comment (PaulBredl) [inline src/main/resources/graphql/query.graphqls:10]: wir brauchen auch irgendwann ne non throwing query

File: src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    @Autowired
+    private FlashcardSideRepository flashcardSideRepository;
+    @Autowired
+    private FlashcardRepository flashcardRepository;
+
+    @Autowired
+    private TestUtils testUtils;
+
+    @Test
+    @Transactional
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java:40]: Bitte mit @Commit. Sonst macht Spring nichts in der Datenbank und man findet keine DB Fehler wie constraint violations

File: src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+        assertThat(flashcardSetRepository.count()).isEqualTo(1);
+        assertThat(flashcardSetRepository.findById(setToDelete)).isEmpty();
+
+        assertThat(flashcardRepository.count()).isEqualTo(2);
+        assertThat(flashcardSideRepository.count()).isEqualTo(4);
+    }
+
+    @Test
+    @Transactional
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java:69]: Bei Tests, die Fehler erwarten allerdings ist das so richtig (@Commit würde hier einen Fehlschlag liefern, weil es ja keinen Commit in der DB gab)

File: src/test/java/de/unistuttgart/iste/gits/flashcard_service/test_utils/TestUtils.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+@Component
+public class TestUtils {
+    /**
+     * Helper method which creates some flashcard sets and saves them to the repository.
+     * @param repo The repository to save the flashcard sets to.
+     * @return Returns the created flashcard sets.
+     */
+    @Transactional(Transactional.TxType.REQUIRES_NEW) // Required so the data is committed to the repo before the
+    // rest of the test method which calls this method is executed.
+    public List<FlashcardSetEntity> populateFlashcardSetRepository(FlashcardSetRepository repo) {
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/flashcard_service/test_utils/TestUtils.java:25]: Interesting way to create test data. I personally prefer builders, but this works too 👍

**Review by PaulBredl — CHANGES_REQUESTED** [review]
minor issue

File: src/main/resources/graphql/query.graphqls (hunk 1 above)
Comment (PaulBredl) [reply src/main/resources/graphql/query.graphqls:10]: (aber nicht jetzt)

File: src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java (hunk 2 above)
Comment (myluki2000) [reply src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java:40]: Ah da hab ich die @Commit annotation falsch verstanden. Hab gedacht die ist dafür da wenn man will dass die Daten von dem Test in die DB geschrieben werden sodass sie dann in einem anderen Test noch da sind

das ist dann in anderen services aber glaub auch oft falsch. müsste man sich mal anschauen

Comment (PaulBredl) [reply src/test/java/de/unistuttgart/iste/gits/flashcard_service/api/MutationDeleteFlashcardSetTest.java:40]: ja kann sein. Aber ist auch kein Weltuntergang in den meisten Fällen

**Review by PaulBredl — APPROVED** [review]

**Pull request merged**